src/crypto/tls/handshake_client_tls13.go

```diff
@@ -535,10 +559,19 @@ func (hs *clientHandshakeStateTLS13) processServerHello() error {
 func (hs *clientHandshakeStateTLS13) establishHandshakeKeys() error {
 	c := hs.c
 
-	sharedKey := hs.ecdheParams.SharedKey(hs.serverHello.serverShare.data)
-	if sharedKey == nil {
-		c.sendAlert(alertIllegalParameter)
-		return errors.New("tls: invalid server key share")
+	var sharedKey []byte
+	if params, ok := hs.keySharePrivate.(ecdheParameters); ok {
+		sharedKey = params.SharedKey(hs.serverHello.serverShare.data)
+	} else if sk, ok := hs.keySharePrivate.(circlKem.PrivateKey); ok {
+		var err error
+		sharedKey, err = sk.Scheme().Decapsulate(sk, hs.serverHello.serverShare.data)
+		if err != nil {
+			c.sendAlert(alertInternalError)
+			return err
+		}
+	} else {
+		c.sendAlert(alertInternalError)
+		return fmt.Errorf("tls: failed to establish handshake keys")
```

Comment:
I think the previous error handling is more appropriate:
```suggestion
		c.sendAlert(alertIllegalParameter)
		return errors.New("tls: invalid server key share")
```

Reply:
Agree. Done.